staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -206,6 +206,10 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
 			// Attempt to gather list in chunks, if supported by listerWatcher, if not, the first
 			// list request will return the full response.
 			pager := pager.New(pager.SimplePageFunc(func(opts metav1.ListOptions) (runtime.Object, error) {
+				if options.ResourceVersion != "" {
+					// FIXME: Remove the limit to force listing from cache.
+					opts.Limit = 0
+				}
```

Comment:
@liggitt - this version is obviously not in mergable shape now, but I think it eliminates the problems that we were discussing yesterday, as we modify clients, not server.

Reply:
Thanks for jumping on this while I was out @wojtek-t.

This looks right. It uses the "Not older than" List semantic instead of the "Exact" list RV semantic (as we documented https://kubernetes.io/docs/reference/using-api/api-concepts/#the-resourceversion-parameter), which should not result in "410 Gone" after api-servers are restarted that is causing the requests to fallback to RV="" and the thundering herd.

When the watch cache is enabled, this seems fine since the watch cache doesn't paginate anyway.

When the watch cache disabled. The relist will be unpaginated. In 1.16 the relist requests were paginated. It's not yet clear to me if that is acceptable or not (cc @smarterclayton). If it is unacceptable, then we might need to rollback https://github.com/kubernetes/kubernetes/pull/83520 and regroup, which I hate to suggest since this PR is otherwise fixes forward the issue quite well.

Reply:
As an aside, this highlights why having the semantics of List be different when a limit is set ("Not older than" if limit=0 and "Exact" if limit > 0) is problematic. It's difficult to reason about.

Oh, and it's really bad when ListerWatcher are added to the mix, which can happen with the Reflectors here in some cases. When the ListerWatcher adds pagination, even if the Reflector makes a limit=0 request, the ListerWatcher might set limit=n and change the semantic.

Reply:
> When the watch cache disabled, the relist will be unpaginated. In 1.16 the relist requests were paginated, so this is a change in behavior. It's not yet clear to me if that is acceptable or not (cc @smarterclayton). If it is unacceptable, then we might need to rollback #83520 and regroup, which I hate to suggest since this PR is otherwise fixes forward the issue quite well.

Yeah - I agree, that was also my slight concern initially.
The counterargument is that I very strongly claim that you can't have very large clusters without having watchcache enabled (at least for the main resources like Nodes, Endpoints, Pods).
But other than this solution, I don't see any other short-term mitigation.


> Oh, and it's really bad when ListerWatcher are added to the mix, which can happen with the Reflectors here in some cases. When the ListerWatcher adds pagination, even if the Reflector makes a limit=0 request, the ListerWatcher might set limit=n and change the semantic.

I also changed the ListWatcher in this PR - currently when reflector already does the pagination, ListWatcher shouldn't do that.

Reply:
BTW - I thought about ListWatcher and this is actually a way how user can enforce pagination. If they will be setting it explicitly in its List function, they can force pagination. But by default, we set it to unpaginated calls.
So maybe that's actually how we should advertise it - if you really want, you can do that via ListWatcher.

Reply:
> The counterargument is that I very strongly claim that you can't have very large clusters without having watchcache enabled (at least for the main resources like Nodes, Endpoints, Pods).

The watch cache uses a reflector to populate its own internal store. Disabling paging here affects that reflector as well, and means that a kube-apiserver could fail to populate its own watch cache for a resource with sufficiently large object counts. xref https://github.com/kubernetes/kubernetes/issues/49684

Reply:
Great cacht.

Fixing watch cache is easy - we can built-in pagination directly into its own ListerWatcher:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L1088

The question is whether there are more thing that it breaks...

Reply:
Another alternative would be to allow for configuring paging function in the reflector (using some setter). Or exposing "always enable pagination" option or sth like that.

Reply:
Yeah... there are options we can explore. That said, this is getting complex enough that I would (sadly) recommend reverting the client changes for 1.17.x and proceed fixing this in master.

Reply:
Yeah - as discussed offline, I (also sadly) tend to agree that it's the right thing to do.
I will continue working on this PR and ask @jpbetz to make a revert in 1.17 branch.

Reply:
I agree. I'll open the revert patch now.

Reply:
Opened https://github.com/kubernetes/kubernetes/pull/86791 to revert.

Once we've got that backported to 1.17, I think we can rebase this fix on master, re-enabled the reflector change and target this fix for 1.18.

Reply:
I went through the codebase (in kubernetes org) to see how reflector is used (there are only ~25 occurrences of it) and in my opinion the only one that may suffer from lack of pagination at the first place was watchcache.
So I introduced the forcePagination option and used it there.

PTAL